src/systems/filecoin_mining/storage_mining/storage_miner_actor.go

```diff
@@ -27,41 +43,50 @@ func (cs *ChallengeStatus_I) OnChallengeResponse(currEpoch block.ChainEpoch) Cha
 	return cs
 }
 
-func (cs *ChallengeStatus_I) isChallenged() bool {
+func (cs *ChallengeStatus_I) IsChallenged() bool {
 	// true (isChallenged) when LastChallengeEpoch is later than LastChallengeEndEpoch
 	return cs.LastChallengeEpoch() > cs.LastChallengeEndEpoch()
 }
 
+func (st *StorageMinerActorState_I) _isChallenged(rt Runtime) bool {
+	return st.ChallengeStatus().IsChallenged()
+}
+
+func (a *StorageMinerActorCode_I) _isChallenged(rt Runtime) bool {
+	h, st := a.State(rt)
+	ret := st._isChallenged(rt)
+	h.Release(st)
+	return ret
+}
+
 // called by CronActor to notify StorageMiner of PoSt Challenge
-func (sm *StorageMinerActor_I) NotifyOfPoStChallenge() {
+func (a *StorageMinerActorCode_I) NotifyOfPoStChallenge(rt Runtime) {
+	rt.ValidateCallerIs(rt.CronActorAddress())
 
-	if sm.ChallengeStatus().Impl().isChallenged() {
+	if a._isChallenged(rt) {
 		return // silent return, dont re-challenge
 	}
 
-	// TODO: get currEpoch from runtime
-	var currEpoch block.ChainEpoch
-	sm.ChallengeStatus().Impl().OnNewChallenge(currEpoch)
-
-	// TODO: commit state change
+	h, st := a.State(rt)
+	st.ChallengeStatus().Impl().LastChallengeEpoch_ = rt.CurrEpoch()
+	h.Commit(st)
 }
 
-func (sm *StorageMinerActor_I) processStagedCommittedSectors() {
-	for sectorNo, sealOnChainInfo := range sm.StagedCommittedSectors() {
-		sm.Sectors()[sectorNo] = sealOnChainInfo
-		sm.ProvingSet_.Add(sectorNo)
-		sm.SectorTable().Impl().CommittedSectors_ += 1
+func (st *StorageMinerActorState_I) _updateProcessStagedCommittedSectors(rt Runtime) {
+	for sectorNo, sealOnChainInfo := range st.StagedCommittedSectors() {
+		st.Sectors()[sectorNo] = sealOnChainInfo
+		st.Impl().ProvingSet_.Add(sectorNo)
```

Comment:
can this just be `st.ProvingSet().Add(sectorNo)` ? or no?

Reply:
Unfortunately, no:

```
systems/filecoin_mining/storage_mining/storage_miner_actor.go:78:18: cannot call pointer method on st.ProvingSet()
systems/filecoin_mining/storage_mining/storage_miner_actor.go:78:18: cannot take the address of st.ProvingSet()
```

I believe the problem is that `ProvingSet()` returns the set by value. We could fix this by implementing functional-style "With" methods in the id compilation, e.g., `With_ProvingSet(newProvingSet)`, but this seems lower priority.